client_plugin/drivers/shared/etcdops.go

```diff
@@ -0,0 +1,271 @@
+func checkManagerAddr(node swarm.Node) (string, error) {
+	if node.Status.State != swarm.NodeStateReady {
+		log.WithFields(log.Fields{"nodeID": node.ID,
+			"state": node.Status.State}).Error("State of node is not ready ")
+		return "", fmt.Errorf("Node %s status %v", node.ID, node.Status.State)
+	}
+
+	// ManagerStatus.Addr is in the format of "host:port", return host only
+	s := strings.Split(node.ManagerStatus.Addr, ":")
+
+	return s[0], nil
+}
+
+// startEtcdCluster function is called by swarm leader to start a ETCD cluster
+func startEtcdCluster(nodeAddr string, nodeID string) error {
+	lines := []string{
+		"--name", nodeID,
+		"--advertise-client-urls", etcdScheme + nodeAddr + etcdClientPort,
+		"--initial-advertise-peer-urls", etcdScheme + nodeAddr + etcdPeerPort,
+		"--listen-client-urls", etcdScheme + etcdListenURL + etcdClientPort,
+		"--listen-peer-urls", etcdScheme + etcdListenURL + etcdPeerPort,
+		"--initial-cluster-token", etcdClusterToken,
+		"--initial-cluster", nodeID + "=" + etcdScheme + nodeAddr + etcdPeerPort,
+		"--initial-cluster-state", etcdClusterStateNew,
+	}
+	go etcdRoutine(lines)
+
+	return nil
+}
+
+// joinEtcdCluster function is called by a non-leader swarm manager to join a ETCD cluster
+func joinEtcdCluster(nodeAddr string, leaderAddr string, nodeID string) error {
+	endpoint := leaderAddr + etcdClientPort
+	cfg := etcdClient.Config{
+		Endpoints: []string{endpoint},
+	}
+
+	cli, err := etcdClient.New(cfg)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"endpoint": endpoint,
+				"error": err},
+		).Error("Failed to create ETCD Client ")
+		return err
+	}
+	defer cli.Close()
+
+	// list all current ETCD members, check if this node is already added as a member
+	lresp, err := cli.MemberList(context.Background())
+	if err != nil {
+		log.WithFields(
+			log.Fields{"endpoint": endpoint,
+				"error":       err,
+				"members len": len(lresp.Members)},
+		).Error("Failed to list member for ETCD")
+		return err
+	}
+
+	peerAddr := etcdScheme + nodeAddr + etcdPeerPort
+	existing := false
+	for _, member := range lresp.Members {
+		if member.PeerURLs[0] == peerAddr {
+			if member.Name == "" {
+				// already added, not started
+				log.WithFields(
+					log.Fields{"nodeID": nodeID,
+						"peerAddr": peerAddr},
+				).Info("This node is already added as a etcd member, but not started yet ")
+
+				existing = true
+			} else {
+				// already added and started, need to reset
+				log.WithFields(
+					log.Fields{"nodeID": nodeID,
+						"peerAddr": peerAddr},
+				).Info("This node is already added as a etcd member and started, need to remove first ")
+
+				_, err = cli.MemberRemove(context.Background(), member.ID)
+				if err != nil {
+					log.WithFields(
+						log.Fields{"peerAddr": peerAddr,
+							"member.ID": member.ID},
+					).Error("Failed to remove member for ETCD")
+					return err
+				}
+			}
+			break
+		}
+	}
+
+	initCluster := ""
+	if !existing {
+		peerAddrs := []string{peerAddr}
```

Comment:
https://github.com/coreos/etcd/blob/d51d381eca406156fd7e44118d4b0dab03589354/clientv3/cluster.go#L36
It's the right name in ETCD world...
It especially means the URL addresses for listening peer traffic.